### PR TITLE
[WIP] tensorflow: fix compatibility issue with tf2 keras

### PR DESCRIPTION
**Pull request by pleasantrabbit**
fix compatibility issue with tf2 keras.

Signed-off-by: Yulu Jia <yulu.jia@bytedance.com>

### CODE REVIEW

**Comment by bobzhuyb** [comment]
Let me know when it is ready to be reviewed.

File: byteps/_keras/callbacks.py [hunk 1]
```diff
@@ -54,13 +67,17 @@ def _average_metrics_in_place(self, logs):
         # Reduce every metric among workers. Sort metrics by name
         # to ensure consistent order.
         for metric, value in sorted(logs.items()):
-            if metric not in self.variables:
-                self.variables[metric], self.push_pull_ops[metric] = \
-                    self._make_variable(metric, value)
+            if bps._executing_eagerly():
+                reduced_logs[metric] = \
+                    bps.allreduce(self.backend.constant(value, name=metric)).numpy()
```
Comment (ymjiang) [inline byteps/_keras/callbacks.py:72]: I don't think we have allreduce api. Should it be `bps.push_pull`?

File: byteps/tensorflow/__init__.py [hunk 2]
```diff
@@ -267,17 +274,77 @@ def get_slot_names(self, *args, **kwargs):
...
+      compression:
+        Compression algorithm used during push_pull to reduce the amount
+        of data sent during each parameter update step.  Defaults to
+        not using compression.
+      sparse_as_dense:
+        Treat all sparse gradients as dense tensors.  This can help improve
+        performance and memory utilization if the original sparse gradient
+        has high density.  Defaults to false.
+      backward_passes_per_step:
+        Number of backward passes to perform before calling hvd.allreduce.
```
Comment (ymjiang) [inline byteps/tensorflow/__init__.py:310]: Should be `bps.push_pull`

File: byteps/_keras/callbacks.py (hunk 1 above)
Comment (pleasantrabbit) [reply byteps/_keras/callbacks.py:72]: right, it should be push_pull

File: byteps/_keras/__init__.py [hunk 3]
```diff
@@ -83,7 +83,7 @@ def apply_gradients(self, *args, **kwargs):
 
 
 def _eval(backend, op_or_result):
-    if hvd._executing_eagerly():
+    if pbs._executing_eagerly():
```
Comment (bobzhuyb) [inline byteps/_keras/__init__.py:86]: pbs -> bps?

File: byteps/_keras/callbacks.py [hunk 4]
```diff
@@ -14,6 +14,8 @@
 # limitations under the License.
 # ==============================================================================
 
+import warnings
```
Comment (bobzhuyb) [inline byteps/_keras/callbacks.py:17]: Where is `warning` used?

File: byteps/tensorflow/__init__.py [hunk 5]
```diff
@@ -16,26 +16,35 @@
 # ==============================================================================
 # pylint: disable=g-short-docstring-punctuation
 
-from __future__ import absolute_import
```
Comment (bobzhuyb) [inline byteps/tensorflow/__init__.py:19]: Why do we remove these __future__ imports? I don't think we are dropping support for python2

File: byteps/tensorflow/__init__.py [hunk 6]
```diff
@@ -267,17 +273,77 @@ def get_slot_names(self, *args, **kwargs):
         def variables(self, *args, **kwargs):
             """Calls this same method on the underlying optimizer."""
             return self._optimizer.variables(*args, **kwargs)
+#x2682marklinexxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```
Comment (bobzhuyb) [inline byteps/tensorflow/__init__.py:276]: #x2682marklinexxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

What is this?

File: byteps/tensorflow/__init__.py [hunk 7]
```diff
@@ -16,26 +16,35 @@
...
+import os
+import warnings
 
 from byteps.tensorflow.compression import Compression
 from byteps.tensorflow.ops import broadcast, _push_pull
 from byteps.tensorflow.ops import init, shutdown, suspend, resume
 from byteps.tensorflow.ops import size, local_size, rank, local_rank
+from byteps.tensorflow.ops import handle_average_backwards_compatibility
 from byteps.tensorflow.util import _executing_eagerly
+# from horovod.tensorflow.util import _executing_eagerly, _make_subgraph, _cache
```
Comment (bobzhuyb) [inline byteps/tensorflow/__init__.py:28]: Clean this

File: byteps/tensorflow/__init__.py (hunk 5 above)
Comment (pleasantrabbit) [reply byteps/tensorflow/__init__.py:19]: I will add it back.

File: byteps/tensorflow/__init__.py (hunk 6 above)
Comment (pleasantrabbit) [reply byteps/tensorflow/__init__.py:276]: this was to make my diff tool behave. will remove.

File: byteps/tensorflow/__init__.py (hunk 7 above)
Comment (pleasantrabbit) [reply byteps/tensorflow/__init__.py:28]: will do.

**Pull request merged**